In [13]:
import pandas as pd
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
import seaborn as sns
import statsmodels.api as sm
from sklearn.preprocessing import PolynomialFeatures
import scipy.stats as stats

file = 'Colesterol.csv'
df = pd.read_csv(file)
df

Total   HDL  Idade
0   177.0  69.0     25
1   249.0  49.0     37
2   167.0  20.0     40
3   202.0  59.0     23
4   186.0  74.0     25
5     NaN  74.0     25
6     NaN  45.0     36
7     NaN  51.0     31
8     NaN  46.0     27
9     NaN  11.0     44
10  156.0   NaN     25
11  274.0   NaN     44
12  260.0   NaN     37
13  316.0   NaN     45
14  320.0   NaN     33
15  211.0  15.0     25
16  174.0  48.0     21
17  198.0  55.0     21
18  261.0  75.0     45
19  250.0  58.0     34
20  452.0  18.0     22
21  279.0  62.0     48
22  235.0  42.0     48
23  310.0  64.0     40
24  167.0  42.0     26
25  179.0  67.0     29
26  203.0  62.0     32
27  293.0  71.0     55
28  189.0  59.0     22

In [14]:
df.isna().sum()

Total    5
HDL      5
Idade    0
dtype: int64

In [15]:
df1 = df
df_filtered = df1.query('Total.notnull() and HDL.notnull() and Idade.notnull()')
df_filtered

Total   HDL  Idade
0   177.0  69.0     25
1   249.0  49.0     37
2   167.0  20.0     40
3   202.0  59.0     23
4   186.0  74.0     25
15  211.0  15.0     25
16  174.0  48.0     21
17  198.0  55.0     21
18  261.0  75.0     45
19  250.0  58.0     34
20  452.0  18.0     22
21  279.0  62.0     48
22  235.0  42.0     48
23  310.0  64.0     40
24  167.0  42.0     26
25  179.0  67.0     29
26  203.0  62.0     32
27  293.0  71.0     55
28  189.0  59.0     22

In [16]:
df_filtered.isna().sum()

Total    0
HDL      0
Idade    0
dtype: int64

In [17]:
x = df_filtered[['Total', 'HDL']]
y = df_filtered['Idade'].to_numpy()
model = LinearRegression()
model.fit(x, y)
score = model.score(x, y)
print(score)

0.15171980214127057


In [18]:
# Vou começar a reconstruir o HDL
# Vou utilizar como variável regressora o Total

df_filtered = df1[ (df1['Total'].notnull()) & (df1['HDL'].notnull()) & (df1['Idade'].notnull()) ]
x = df_filtered[['Total']].to_numpy()
y = df_filtered['HDL'].to_numpy()
model_hdl = LinearRegression()
model_hdl.fit(x, y)
score = model_hdl.score(x, y)
score

0.031590450080245525

In [19]:
def completar_hdl(linha):
    if pd.isna(linha['HDL']) and not pd.isna(linha['Total']) and not pd.isna(linha['Idade']):
        return round(model_hdl.predict([[linha['Total']]])[0])
    return linha['HDL']

# apply: executar a função chamada 1x para cada linha
df1['HDL_calculado'] = df1.apply(completar_hdl, axis=1)
df1

Total   HDL  Idade  HDL_calculado
0   177.0  69.0     25           69.0
1   249.0  49.0     37           49.0
2   167.0  20.0     40           20.0
3   202.0  59.0     23           59.0
4   186.0  74.0     25           74.0
5     NaN  74.0     25           74.0
6     NaN  45.0     36           45.0
7     NaN  51.0     31           51.0
8     NaN  46.0     27           46.0
9     NaN  11.0     44           11.0
10  156.0   NaN     25           57.0
11  274.0   NaN     44           51.0
12  260.0   NaN     37           52.0
13  316.0   NaN     45           49.0
14  320.0   NaN     33           49.0
15  211.0  15.0     25           15.0
16  174.0  48.0     21           48.0
17  198.0  55.0     21           55.0
18  261.0  75.0     45           75.0
19  250.0  58.0     34           58.0
20  452.0  18.0     22           18.0
21  279.0  62.0     48           62.0
22  235.0  42.0     48           42.0
23  310.0  64.0     40           64.0
24  167.0  42.0     26           42.0
25  179.0  67.0     29           67.0
26  203.0  62.0     32           62.0
27  293.0  71.0     55           71.0
28  189.0  59.0     22           59.0

In [20]:
df_filtered = df1[ (df1['Total'].notnull()) & (df1['Idade'].notnull()) & (df1['HDL_calculado'].notnull()) ]
x = df_filtered[['HDL_calculado']].to_numpy()
y = df_filtered['Total'].to_numpy()
model_total = LinearRegression()
model_total.fit(x, y)
score = model_total.score(x, y)
score

0.039732895242611255

In [21]:
def completar_total(linha):
    if pd.isna(linha['Total']) and not pd.isna(linha['Idade']) and not pd.isna(linha['HDL_calculado']):
        return round(model_total.predict([[linha['HDL_calculado']]])[0])
    return linha['Total']

# apply: executar a função chamada 1x para cada linha
df1['Total_calculado'] = df1.apply(completar_total, axis=1)
df1

Total   HDL  Idade  HDL_calculado  Total_calculado
0   177.0  69.0     25           69.0            177.0
1   249.0  49.0     37           49.0            249.0
2   167.0  20.0     40           20.0            167.0
3   202.0  59.0     23           59.0            202.0
4   186.0  74.0     25           74.0            186.0
5     NaN  74.0     25           74.0            220.0
6     NaN  45.0     36           45.0            244.0
7     NaN  51.0     31           51.0            239.0
8     NaN  46.0     27           46.0            244.0
9     NaN  11.0     44           11.0            273.0
10  156.0   NaN     25           57.0            156.0
11  274.0   NaN     44           51.0            274.0
12  260.0   NaN     37           52.0            260.0
13  316.0   NaN     45           49.0            316.0
14  320.0   NaN     33           49.0            320.0
15  211.0  15.0     25           15.0            211.0
16  174.0  48.0     21           48.0            174.0
17  198.0  55.0     21           55.0            198.0
18  261.0  75.0     45           75.0            261.0
19  250.0  58.0     34           58.0            250.0
20  452.0  18.0     22           18.0            452.0
21  279.0  62.0     48           62.0            279.0
22  235.0  42.0     48           42.0            235.0
23  310.0  64.0     40           64.0            310.0
24  167.0  42.0     26           42.0            167.0
25  179.0  67.0     29           67.0            179.0
26  203.0  62.0     32           62.0            203.0
27  293.0  71.0     55           71.0            293.0
28  189.0  59.0     22           59.0            189.0

In [22]:
# Recalculo a minha regressão original utilizando as variáveis novas
x = df1[['Total_calculado', 'HDL_calculado']]
y = df1['Idade'].to_numpy()
model = LinearRegression()
model.fit(x, y)
score = model.score(x, y)
print(score)

0.14931397313959738


* A regressão Linear sem o tratamento dos missings, obteve um r^2 de 0.15
* A diferença percentual no poder de explicação entre os dois coeficientes de determinação é aproximadamente 1.58%
* Já para regressão com os missings tratados, obteve uma significância menor com um r^2 de 0.14, mas sem viés e um poder de explicação maior